app/controllers/api/v1/courses_controller.rb

```diff
@@ -0,0 +1,120 @@
+include ActionController::Flash
+
+class Api::V1::CoursesController < ApplicationController
+  before_action :set_course, only: %i[ show update destroy ]
+
+  # GET /courses
+  # List all the courses
+  def index
+    @courses = Course.all
+
+    render json: @courses
+  end
+
+  # GET /courses/1
+  # Get a course
+  def show
+    render json: @course
+  end
+
+  # POST /courses
+  # Create a course
+  def create
+    @course = Course.new(course_params)
+
+    if @course.save
+      render json: @course, status: :created
+    else
+      render json: @course.errors, status: :unprocessable_entity
+    end
+  end
+
+  # PATCH/PUT /courses/1
+  # Update a course
+  def update
+    if @course.update(course_params)
+      render json: @course
+    else
+      render json: @course.errors, status: :unprocessable_entity
+    end
+  end
+
+  # DELETE /courses/1
+  # Delete a course
+  def destroy
+    @course.destroy
+    render json: { message: "Course with id #{params[:id]}, deleted" }, status: :ok
+  end
+
+  # Adds a Teaching Assistant to the course
+  def add_ta
+    @course = Course.find(params[:id])
+    @user = User.find_by(id: params[:ta_id])
+    if @user.nil?
+      return render json: { status: "error", message: "The user " + params[:ta_id].to_s + " does not exist" }, status: :bad_request
+    elsif !TaMapping.where(ta_id: @user.id, course_id: @course.id).empty?
+      return render json: { status: "error", message: "The user " + params[:ta_id].to_s + " is already a TA for this course." }, status: :bad_request
+    else
+      @ta_mapping = TaMapping.create(ta_id: @user.id, course_id: @course.id)
+      @role_id = Role.find_by(name: 'Teaching Assistant').id
+      @user.update_attribute(:role_id, @role_id)
+      @user.save
+    end
+
+    if @ta_mapping.save
+      render json: @ta_mapping.slice(:course_id, :ta_id), status: :created
+    else
+      render json: @ta_mapping.errors, status: :unprocessable_entity
+    end
+  end
+
+  # Displays all Teaching Assistants for the course
+  def view_tas
+    @course = Course.find(params[:id])
+    @ta_mappings = @course.ta_mappings
+    @users = User.where(id: @ta_mappings.pluck(:ta_id))
+    render json: @users, status: :ok
+  end
+
+  # Removes Teaching Assistant from the course
+  def remove_ta
```

Comment:
Again, logic needs to be in model
here's the possible refactoring

```ruby
class Course < ApplicationRecord
  def remove_ta(ta_id)
    ta_mapping = ta_mappings.find_by(ta_id: ta_id)

    return { success: false, message: "No TA mapping found for the specified course and TA" } if ta_mapping.nil?

    ta = User.find(ta_mapping.ta_id)
    ta_count = TaMapping.where(ta_id: ta_id).size - 1

    if ta_count.zero?
      ta.update(role: Role::STUDENT)
    end

    ta_mapping.destroy
    { success: true, ta_name: ta.name }
  end
end
```

In the controller:
```ruby
def remove_ta
  @course = Course.find(params[:id])
  result = @course.remove_ta(params[:ta_id])

  if result[:success]
    render json: { message: "The TA #{result[:ta_name]} has been removed." }, status: :ok
  else
    render json: { status: "error", message: result[:message] }, status: :not_found
  end
end
```

Reply:
Thanks for the suggestion; I will add the logic to the model.